kiwi_stackbuild_plugin/tasks/system_stash.py

```diff
@@ -0,0 +1,128 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with kiwi.  If not, see <http://www.gnu.org/licenses/>
+#
+"""
+usage: kiwi-ng system stash -h | --help
+       kiwi-ng system stash --root=<directory> --target-dir=<directory>
+           [--tag=<name>]
+           [--maintainer=<name>]
+           [--name=<container_name>]
+       kiwi-ng system stash help
+
+commands:
+    stash
+        create a container from the given root directory
+    stash help
+        show manual page for stash command
+
+options:
+    --root=<directory>
+        the path to the root directory, usually the result of
+        a former system prepare or build call
+    --target-dir=<directory>
+        the target directory to store the container files
+    --tag=<name>
+        the tag name for the container.
+        By default set to: latest
+    --maintainer=<name>
+        the maintainer name used in the history/author and label
+        section of the container metadata.
+        By default set to: KIWI
+    --name=<container_name>
+        the name of the container as it appears in the registry.
+        By default set to: kiwi-stash-DATE
+"""
+import os
+import logging
+from datetime import datetime
+from typing import Dict
+
+from kiwi.tasks.base import CliTask
+from kiwi.help import Help
+from kiwi.defaults import Defaults
+from kiwi.privileges import Privileges
+from kiwi.container import ContainerImage
+from kiwi.container.setup import ContainerSetup
+from kiwi.path import Path
+
+
+log = logging.getLogger('kiwi')
+
+
+class SystemStashTask(CliTask):
+    def process(self) -> None:
+        self.manual = Help()
+        if self.command_args.get('help') is True:
+            return self.manual.show('kiwi::system::stash')
+
+        Privileges.check_for_root_permissions()
+
+        abs_target_dir_path = os.path.abspath(
+            self.command_args['--target-dir']
+        )
+        Path.create(abs_target_dir_path)
+
+        container_name = self.command_args['--name'] or 'kiwi-stash-{0}'.format(
+            datetime.utcnow().strftime('%Y%m%dt%H%M%S')
+        )
+        container_file = ''.join(
+            [
+                abs_target_dir_path, '/', container_name,
+                '.' + Defaults.get_platform_name(),
+                '.docker.tar'
+            ]
+        )
+        container_config = SystemStashTask._get_container_config(
+            container_name,
+            self.command_args['--tag'],
+            self.command_args['--maintainer']
+        )
+
+        container_setup = ContainerSetup.new(
+            'docker', self.command_args['--root'], container_config
+        )
+        container_setup.setup()
+        container_image = ContainerImage.new(
+            'docker', self.command_args['--root'], container_config
```

Comment:
We are facing the weaknesses of the container API in KIWI here. Note that this `docker` here impacts on the `transport` concept of image references of tools such as  `podman` or `skopeo`. Since KIWI only works at file level this is not a problem there. 

See how we loose the flexibility to set a custom transport in the container factory class. `ContainerImageOCI` can be initiated with a custom transport, but not from the factory class. I was not satisfied with this implementation at a time already, but I given we were only dealing with file images I though it was good enough for the time being.

Just to be explicit the _transport_ field of `ContainerImageOCI` class states how to store the resulting image `docker-archive` means a tarball (ready for `docker load`), `containers-storage` puts into a local storage as used by podman or buildah, `docker-daemon` could be used to inject it in a docker daemon storage and I believe a bunch of others are also available (see [skopeo README](https://github.com/containers/skopeo)) including remote registries.

At the end it all translates into a `skopeo copy <kiwi working image> <transport>:<target image reference>`. In KIWI this happens in the `OCI.export_container_image` method, which is called as one of the last steps of the OCI container builder.

So, here I would initiate directly a `ContainerImageOCI` image instance or refactor KIWI to make transport customizable once the `ContainerImageOCI` instance is already initiated form the factory.

I am not sure I managed to explain myself properly :sweat_smile:

Reply:
Yeah it has turned out that I misunderstood parts of the OCI concept but you were able to clarify all of them and in the new implementation I use the OCI interface as you suggested.